src/Build/Evaluation/Expander.cs

```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
         }
 
         /// <summary>
-        /// Add the argument in the StringBuilder to the arguments list, handling nulls
+        /// Add the argument represented by the slices to the arguments list, handling nulls
         /// appropriately.
         /// </summary>
-        private static void AddArgument(List<string> arguments, SpanBasedStringBuilder argumentBuilder)
+        private static void AddArgumentFromSlices(List<string> arguments, List<Tuple<int, int>> slices, string arg)
         {
-            // we reached the end of an argument, add the builder's final result
-            // to our arguments.
-            argumentBuilder.Trim();
-            string argValue = argumentBuilder.ToString();
+            // This shouldn't happen
+            int firstSlice = 0;
+            int lastSlice = slices.Count - 1;
+            if (lastSlice == -1)
+            {
+                arguments.Add("");
+                return;
+            }
+            // from end
+            int lastSliceIdx = slices[slices.Count - 1].Item2;
+            // from start
+            int firstSliceIdx = slices[0].Item1;
 
-            // We support passing of null through the argument constant value null
-            if (String.Equals("null", argValue, StringComparison.OrdinalIgnoreCase))
+            // Trim from the start
+            while (firstSlice <= lastSlice && Char.IsWhiteSpace(arg, firstSliceIdx))
             {
-                arguments.Add(null);
+                firstSliceIdx++;
+                if (firstSliceIdx > slices[firstSlice].Item2 && firstSlice < lastSlice)
+                {
+                    firstSlice++;
+                    firstSliceIdx = slices[firstSlice].Item1;
+                }
             }
-            else
+
+            // Trim from the end.
+            // There is some extra logic here to avoid edge case where we would trim a whitespace only character with
+            // one slice from the start, and then once more from the end, resulting in invalid indices.
+            while (((firstSlice < lastSlice) || (firstSlice == lastSlice && firstSliceIdx < lastSliceIdx)) && Char.IsWhiteSpace(arg, lastSliceIdx - 1))
```

Comment:
it looks like `(firstSlice <= lastSlice && Char.IsWhiteSpace(arg, firstSliceIdx)` can be moved to the separate method with descriptive name and reused in the while loop here and on the line 794

Reply:
This checks checks against lastSliceIndx and also checks to prevent duplicate whitespace removal (the second part after the ||)
It was similar-ish before, then I run into the "all whitespace" edge case and crashed. 
that is what the `firstSlice == lastSlice && firstSliceIdx < lastSliceIdx` is for.
So the function would have to be something like
`IsEdgeWhitespace(arg, firstSlice, lastSlice, idxToCheck)` and then some extra logic for this case.
Is that more readable? I would like to say no, but that could be just my laziness/lack of feeling for cases such as this.